specs/deneb/polynomial-commitments-sampling.md

```diff
@@ -0,0 +1,523 @@
+    """
+    Lagrange interpolation: Finds the lowest degree polynomial that takes the value ``ys[i]`` at ``x[i]``
+    for all i.
+    Outputs a coefficient form polynomial. Leading coefficients may be zero.
+    """
+    assert len(xs) == len(ys)
+    r = [0]
+
+    for i in range(len(xs)):
+        summand = [ys[i]]
+        for j in range(len(ys)):
+            if j != i:
+                weight_adjustment = bls_modular_inverse(int(xs[i]) - int(xs[j]))
+                summand = multiply_polynomialcoeff(
+                    summand, [(- int(weight_adjustment) * int(xs[j])) % BLS_MODULUS, weight_adjustment]
+                )
+        r = add_polynomialcoeff(r, summand)
+    
+    return r
+```
+
+#### `zero_polynomialcoeff`
+
+```python
+def zero_polynomialcoeff(xs: Sequence[BLSFieldElement]) -> PolynomialCoeff:
+    """
+    Compute a zero polynomial on ``xs`` (in coefficient form)
+    """
+    p = [1]
+    for x in xs:
+        p = multiply_polynomialcoeff(p, [-int(x), 1])
+    return p
+```
+
+#### `evaluate_polynomialcoeff`
+
+```python
+def evaluate_polynomialcoeff(polynomial_coeff: PolynomialCoeff, z: BLSFieldElement) -> BLSFieldElement:
+    """
+    Evaluate a coefficient form polynomial at ``z`` using Horner's schema
+    """
+    y = 0
+    for coef in polynomial_coeff[::-1]:
+        y = (int(y) * int(z) + coef) % BLS_MODULUS
+    return BLSFieldElement(y % BLS_MODULUS)
+```
+
+### KZG multiproofs
+
+Extended KZG functions for multiproofs
+
+#### `compute_kzg_proof_multi_impl`
+
+```python
+def compute_kzg_proof_multi_impl(
+        polynomial_coeff: PolynomialCoeff,
+        zs: Sequence[BLSFieldElement]) -> Tuple[KZGProof, Sequence[BLSFieldElement]]:
+    """
+    Helper function that computes multi-evaluation KZG proofs.
+    """
+
+    # For all x_i, compute p(x_i) - p(z)
+    ys = [evaluate_polynomialcoeff(polynomial_coeff, z) for z in zs]
+    interpolation_polynomial = interpolate_polynomialcoeff(zs, ys)
+    polynomial_shifted = add_polynomialcoeff(polynomial_coeff, neg_polynomialcoeff(interpolation_polynomial))
+
+    # For all x_i, compute (x_i - z)
+    denominator_poly = zero_polynomialcoeff(zs)
+
+    # Compute the quotient polynomial directly in evaluation form
+    quotient_polynomial = divide_polynomialcoeff(polynomial_shifted, denominator_poly)
+
+    return KZGProof(g1_lincomb(KZG_SETUP_G1[:len(quotient_polynomial)], quotient_polynomial)), ys
+```
+
+#### `verify_kzg_proof_multi_impl`
+
+```python
+def verify_kzg_proof_multi_impl(commitment: KZGCommitment,
+                                zs: BLSFieldElement,
+                                ys: BLSFieldElement,
+                                proof: KZGProof) -> bool:
+    """
+    Helper function that verifies a KZG multiproof
+    """
+    zero_poly = g2_lincomb(KZG_SETUP_G2[:len(zs) + 1], zero_polynomialcoeff(zs))
+    interpolated_poly = g1_lincomb(KZG_SETUP_G1[:len(zs)], interpolate_polynomialcoeff(zs, ys))
```

Comment:
Double-checking: are all the `KZG_SETUP_G2` are actually `KZG_SETUP_G2_MONOMIAL`, and `KZG_SETUP_G1 ` are `KZG_SETUP_G1_LAGRANGE` in this new doc?

Reply:
Good question! I believe the G1 points there are actually in monomial form. When loading the trusted setup, we will need to do the (simple) conversion when loading the trusted setup. We should rename those to `KZG_SETUP_G1_MONOMIAL` and `KZG_SETUP_G2_MONOMIAL` for clarity.

Reply:
I see!

Hmm, it may make the presets messy again.

We now use the `trusted_setup_4096.json` from the KZG ceremony output, which doesn't have `g1_monomial`. We will need to find a proper place to provide the `KZG_SETUP_G1_MONOMIAL` preset to the client teams.

We can do the conversion in pyspec, but the cleanest solution may be making the KZG ceremony output provide both Lagrange and monomial format for both G1 and G2.

Reply:
We never need monomial for G2.

Reply:
@dankrad 

isn't the [Deneb `KZG_SETUP_G2_MONOMIAL`](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md#trusted-setup) in monomial format? Or it's a naming issue?

Reply:
I think he meant we'll never need G2 points in Lagrange form.

Reply:
@jtraglia is right, edited.

Reply:
Added G1 points in monomial form to the verifier: https://github.com/ethereum/kzg-ceremony-verifier/blob/master/output_setups/trusted_setup_4096.json